### PR TITLE
New C++ API: Add tests for mk-functions in solver object.

### PR DESCRIPTION
**Pull request by aniemetz**
This incorporates the changes of #2674 regarding mkReal/mkInteger.

### CODE REVIEW

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkReal(int64_t num, int64_t den) const
 {
-  return d_exprMgr->mkConst(Rational(num, den));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: `return mk_real_helper(d_ExprMgr, Rational(num, den));`

File: src/api/cvc4cpp.cpp [hunk 2]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkReal(uint32_t num, uint32_t den) const
 {
-  return d_exprMgr->mkConst(Rational(num, den));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: `return mk_real_helper(d_ExprMgr, Rational(num, den));`

File: src/api/cvc4cpp.cpp [hunk 3]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkReal(uint64_t num, uint64_t den) const
 {
-  return d_exprMgr->mkConst(Rational(num, den));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: `return mk_real_helper(d_ExprMgr, Rational(num, den));`

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkReal(int32_t num, int32_t den) const
 {
-  return d_exprMgr->mkConst(Rational(num, den));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: `return mk_real_helper(d_ExprMgr, Rational(num, den));`

File: src/api/cvc4cpp.cpp [hunk 5]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkBitVector(uint32_t size, uint32_t val) const
 {
-  return d_exprMgr->mkConst(BitVector(size, val));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Again something like mk_bitvector_helper?

File: src/api/cvc4cpp.cpp [hunk 6]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkBitVector(uint32_t size, uint64_t val) const
 {
-  return d_exprMgr->mkConst(BitVector(size, val));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Ditto.

File: src/api/cvc4cpp.cpp [hunk 7]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
 }
 
 Term Solver::mkBitVector(const char* s, uint32_t base) const
 {
-  return d_exprMgr->mkConst(BitVector(s, base));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Call `return Solver::mkBitVector(std::string(s), base);`?

File: src/api/cvc4cpp.cpp [hunk 8]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
 }
 
-Term Solver::mkConst(Kind kind, uint64_t arg) const
+Term Solver::mkConst(Kind kind, const char* arg) const
 {
-  CVC4_API_KIND_CHECK_EXPECTED(kind == ABSTRACT_VALUE || kind == CONST_RATIONAL,
-                               kind)
-      << "ABSTRACT_VALUE or CONST_RATIONAL";
-  if (kind == ABSTRACT_VALUE)
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Call `Solver::mkConst(kind, std::string(arg))`?

File: src/api/cvc4cpp.cpp [hunk 9]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+  }
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
+}
+
+Term Solver::mkConst(Kind kind, const char* arg1, uint32_t arg2) const
+{
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Call `return Solver::mkConst(kind, std::string(arg1), arg2)`?

File: src/api/cvc4cpp.cpp [hunk 10]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
+    }
+    return res;
+  }
+  catch (TypeCheckingException& e)
+  {
+    throw CVC4ApiException(e.getMessage());
+  }
+}
+
+Term Solver::mkConst(Kind kind, int32_t arg) const
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Templated helper for `int32_t`, `int64_t`, and `uint64_t`?

File: src/api/cvc4cpp.cpp [hunk 11]
```diff
@@ -1873,348 +1906,868 @@ Term Solver::mkFalse(void) const { return d_exprMgr->mkConst<bool>(false); }
...
   }
-  return d_exprMgr->mkConst(CVC4::Rational(arg));
 }
 
 Term Solver::mkConst(Kind kind, uint32_t arg1, uint32_t arg2) const
 {
-  CVC4_API_KIND_CHECK_EXPECTED(kind == CONST_RATIONAL, kind)
-      << "CONST_RATIONAL";
-  return d_exprMgr->mkConst(CVC4::Rational(arg1, arg2));
+  try
```
Comment (mpreiner) [inline src/api/cvc4cpp.cpp]: Again template helper for `uint32_t`, `int32_t`, `int64_t`, and `uint64_t`?

File: src/api/cvc4cpp.cpp (hunk 10 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp]: Solved it differently, have a look.

File: src/api/cvc4cpp.cpp (hunk 11 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp]: Ditto

**Pull request merged**